test/unittests/execute_floating_point_test.cpp

```diff
@@ -551,6 +551,69 @@ TEST(execute_floating_point, f64_promote_f32)
     EXPECT_GT(FP{res4.value.f64}.nan_payload(), FP64::canon);
 }
 
+TEST(execute_floating_point, f32_demote_f64)
+{
+    /* wat2wasm
+    (func (param f64) (result f32)
+      local.get 0
+      f32.demote_f64
+    )
+    */
+    const auto wasm = from_hex("0061736d0100000001060160017c017d030201000a070105002000b60b");
+    auto instance = instantiate(parse(wasm));
+
+    // The boundary input value that results in the infinity.
+    constexpr double lowest_to_inf = 0x1.ffffffp127;
```

Comment:
I'm a bit confused and might be wrong, but cases `p > lowest_to_inf` and `p < -lowest_to_inf` are missing, and I think they are undefined behavior in C++, but defined to result in infinities in wasm.

Reply:
I think it is defined if IEEE 754 is supported (as for some other cases) because then infinities also counts.

> The minimum range of representable values for a floating type is the most negative finite floating-point number representable in that type through the most positive finite floating-point number representable in that type. In addition, if negative infinity is representable in a type, the range of that type is extended to all negative real numbers; likewise, if positive infinity is representable in a type, the range of that type is extended to all positive real numbers.

https://stackoverflow.com/a/25831946/725174

Although, I'm confused how to chose nearest value between `max` and `infinity`...

The cases you mentions are not present because `lowest_to_inf` already yields infinity, so for any `p > lowest_to_inf` we get infinity too. But I can add test for `nextafter(lowest_to_inf)`.

Reply:
This answers my question.

> In the default IEEE 754 round-to-nearest mode, a few double values above the maximum finite float (that is, FLT_MAX) convert to FLT_MAX. The exact limit is the number midway between FLT_MAX (0x1.fffffep127 in C99 hexadecimal representation) and the next float number that could be represented if the exponent in the single-precision format had a larger range, 0x2.0p127. The limit is thus 0x1.ffffffp127 or approximately 3.4028235677973366e+38 in decimal.

https://stackoverflow.com/a/17751145/725174

Although I wander where wasm spec handles this, if anywhere.

Reply:
I finally have explanation for the `lowest_to_inf` value (see comments).
I also added two cases for `p > lowest_to_inf`.